packages/eslint-plugin/src/rules/no-mixed-enums.ts

```diff
@@ -0,0 +1,222 @@
+import type { Scope } from '@typescript-eslint/scope-manager';
+import { DefinitionType } from '@typescript-eslint/scope-manager';
+import type { TSESTree } from '@typescript-eslint/utils';
+import { AST_NODE_TYPES } from '@typescript-eslint/utils';
+import * as tsutils from 'tsutils';
+import * as ts from 'typescript';
+
+import * as util from '../util';
+
+type AllowedType = ts.TypeFlags.Number | ts.TypeFlags.String;
+
+export default util.createRule({
+  name: 'no-mixed-enums',
+  meta: {
+    docs: {
+      description: 'Disallow enums from having both number and string members',
+      recommended: 'strict',
+      requiresTypeChecking: true,
+    },
+    messages: {
+      mixed: `Mixing number and string enums can be confusing.`,
+    },
+    schema: [],
+    type: 'problem',
+  },
+  defaultOptions: [],
+  create(context) {
+    const parserServices = util.getParserServices(context);
+    const typeChecker = parserServices.program.getTypeChecker();
+
+    interface CollectedDefinitions {
+      imports: TSESTree.Node[];
+      previousSibling: TSESTree.TSEnumDeclaration | undefined;
+    }
+
+    function collectNodeDefinitions(
+      node: TSESTree.TSEnumDeclaration,
+    ): CollectedDefinitions {
+      const { name } = node.id;
+      const found: CollectedDefinitions = {
+        imports: [],
+        previousSibling: undefined,
+      };
+      let scope: Scope | null = context.getScope();
+
+      for (const definition of scope.upper?.set.get(name)?.defs ?? []) {
+        if (
+          definition.node.type === AST_NODE_TYPES.TSEnumDeclaration &&
+          definition.node.range[0] < node.range[0] &&
+          definition.node.members.length > 0
+        ) {
+          found.previousSibling = definition.node;
+          break;
+        }
+      }
+
+      while (scope) {
+        scope.set.get(name)?.defs?.forEach(definition => {
+          if (definition.type === DefinitionType.ImportBinding) {
+            found.imports.push(definition.node);
+          }
+        });
+
+        scope = scope.upper;
+      }
+
+      return found;
+    }
+
+    function getAllowedTypeForNode(node: ts.Node): AllowedType | undefined {
```

Comment:
```suggestion
    function getAllowedTypeForNode(node: ts.Node): AllowedType {
```

nit - never returns `undefined`